[Concurrency] Fix preconcurrency downgrade behavior for `Sendable` closures and generic requirements.

hborla opened the pull request:
Sendable violations inside `@preconcurrency @Sendable` closures should be suppressed in minimal checking, and diagnosed as warnings under complete checking, including the Swift 6 language mode. This change fixes an issue where warnings were still produced in minimal checking, and errors were still produced under `-swift-version 6`, which means that libraries cannot stage in new `Sendable` annotations if their clients have already migrated to Swift 6.

The preconcurrency downgrade infrastructure is currently pretty difficult to follow, because the information is passed through `SendableCheckContext` throughout the `diagnoseNonSendableTypes` APIs, and there are different code paths for `@preconcurrency` on nominal declarations versus import statements. This PR does not attempt to clean any of this code up, because I'd like to cherry pick the bug fix to `release/6.0`. I'll look at clarifying this code in a follow up change.

This change also downgrades `Sendable` requirement failures when the requirement is on a `@preconcurrency` declaration. However, the change in this PR is not sufficient for requirement inference, e.g.

```swift

// module A
@preconcurrency
struct RequireSendable<T: Sendable> {}

// module B

class NotSendable {}

func f<T, U: Sendable>(_: T, _: U) -> RequireSendable<T>? { nil }

func call(ns: NotSendable) {
  f(ns, ns)
}
```

A case like the above requires more careful handling. `@preconcurrency` can't simply be inferred on `f`,  because that would break ABI by stripping `U: Sendable` from the generic signature. `@preconcurrency` may need to apply per requirement in order to maintain source compatibility without breaking ABI.

Resolves: rdar://138535438, rdar://139234188, https://github.com/swiftlang/swift/issues/76652

hborla committed:
[Concurrency] Fix preconcurrency downgrade behavior for Sendable closures.

Sendable violations inside `@preconcurrency @Sendable` closures should be
suppressed in minimal checking, and diagnosed as warnings under complete
checking, including the Swift 6 language mode.

hborla committed:
[Concurrency] Downgrade `@preconcurrency @Sendable` conversion failures
to warnings even in Swift 6.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test

hborla committed:
[Concurrency] Downgrade `@preconcurrency` conformance requirement failures to
warnings in Swift 6 mode.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

DougGregor reviewed (approved):
This all looks safe, and will downgrade a number of errors to warnings based on `@preconcurrency`. Thank you!

Lukasa commented:
@swift-ci build toolchain

Lukasa commented:
@swift-ci build toolchain macOS platform

Lukasa commented:
Hmm, this toolchain did not appear to work. Building the current `main` of Hummingbird using the current `main` of swift-nio (by putting swift-nio in edit mode) produces the same errors using the downloaded Ubuntu 20.04 toolchain as we see with the current 6.0 toolchain:

```
/hummingbird/Sources/HummingbirdTLS/TLSChannel.swift:40:44: error: conformance of 'NIOSSLHandler' to 'Sendable' is unavailable
38 |     @inlinable
39 |     public func setup(channel: Channel, logger: Logger) -> EventLoopFuture<Value> {
40 |         return channel.pipeline.addHandler(NIOSSLServerHandler(context: self.sslContext)).flatMap {
   |                                            `- error: conformance of 'NIOSSLHandler' to 'Sendable' is unavailable
41 |             self.baseChannel.setup(channel: channel, logger: logger)
42 |         }

/hummingbird/.build/checkouts/swift-nio-ssl/Sources/NIOSSL/NIOSSLHandler.swift:791:1: note: conformance of 'NIOSSLHandler' to 'Sendable' has been explicitly marked unavailable here
 789 | 
 790 | @available(*, unavailable)
 791 | extension NIOSSLHandler: Sendable {}
     | `- note: conformance of 'NIOSSLHandler' to 'Sendable' has been explicitly marked unavailable here
 792 | 
 793 | extension NIOSSLHandler {
/hummingbird/Sources/HummingbirdTLS/TLSChannel.swift:40:44: error: conformance of 'NIOSSLHandler' to 'Sendable' is unavailable
38 |     @inlinable
39 |     public func setup(channel: Channel, logger: Logger) -> EventLoopFuture<Value> {
40 |         return channel.pipeline.addHandler(NIOSSLServerHandler(context: self.sslContext)).flatMap {
   |                                            `- error: conformance of 'NIOSSLHandler' to 'Sendable' is unavailable
41 |             self.baseChannel.setup(channel: channel, logger: logger)
42 |         }

/hummingbird/.build/checkouts/swift-nio-ssl/Sources/NIOSSL/NIOSSLHandler.swift:791:1: note: conformance of 'NIOSSLHandler' to 'Sendable' has been explicitly marked unavailable here
 789 | 
 790 | @available(*, unavailable)
 791 | extension NIOSSLHandler: Sendable {}
     | `- note: conformance of 'NIOSSLHandler' to 'Sendable' has been explicitly marked unavailable here
 792 | 
 793 | extension NIOSSLHandler {
```

This is the [`addHandler` method](https://github.com/apple/swift-nio/blob/6d30ec4738389e875f1a22345463468d4d581063/Sources/NIOCore/ChannelPipeline.swift#L170-L175) in question, with the `@preconcurrency` annotation on it.

hborla commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -6214,7 +6239,20 @@ bool swift::contextRequiresStrictConcurrencyChecking(
         // Don't take any more cues if this only got its type information by
         // being provided to a `@preconcurrency` operation.
+        //
+        // FIXME: contextRequiresStrictConcurrencyChecking is called from
+        // within the constraint system, but closures are only set to be isolated
+        // by preconcurrency in solution application because it's dependent on
+        // overload resolution. The constraint system either needs to check its
+        // own state on the current path, or not make type inference decisions based
+        // on concurrency checking level.
         if (isolatedByPreconcurrency(explicitClosure)) {
```
Yes, and it looks like `ClosureIsolatedByPreconcurrency` (used by constraint system calls) does check the constraint system state, but I ran into a case while debugging where this line of code returned an unexpected result mid constraint solving. It didn't seem to be causing any issues for the preconcurrency downgrade behavior, so I left this comment for myself to figure out what happened there after I land this.

hborla committed:
[Concurrency] Downgrade `Sendable` requirement failures in existential
erasure expressions for arguments to `@preconcurrency` functions in Swift
6 mode.

hborla commented:
@swift-ci please smoke test

hborla committed:
[Concurrency] Downgrade preconcurrency errors about unavailable conformances
to `Sendable`.

hborla commented:
@swift-ci build toolchain

hborla commented:
@swift-ci please smoke test
